### PR TITLE
Feature/issue 950 - /v2/entities/{entityId}

### PR DESCRIPTION
**Pull request by crbrox**
Add GET /v2/entities/[id]

Implements issue #950

@fgalan 
@jmcanterafonseca

### CODE REVIEW

File: src/lib/apiTypesV2/Entity.cpp [hunk 1]
```diff
@@ -128,7 +136,31 @@ void Entity::fill(const std::string& _id, const std::string& _type, const std::s
...
+    // If no entity has been found, we respond with a 404 NOT FOUND
+    // If any other error - use the error for the response
+    //
+    errorCode.fill(qcrsP->errorCode);
+  }
+  else if (qcrsP->contextElementResponseVector.size()>1)
+  {
+      //
+      // If there are more than one entity, we return an error
+      // TODO: determine error for this case
```
Comment (fgalan) [inline src/lib/apiTypesV2/Entity.cpp]: Lefvover comment.

File: test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity.test [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+--SHELL-INIT--
+dbInit CB
+brokerStart CB
+
+--SHELL--
+
+#
+# 01. POST /v1/updateContext
+# 03. GET /v2/entities
```
Comment (fgalan) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity.test]: GET /v2/entities -> GET /v2/entities/8787GHY

File: test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity.test [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+      ]
+    }
+  ],
+  "updateAction": "APPEND"
+}'
+orionCurl --url /v1/updateContext --payload "$payload" --json
+echo
+echo
+
+echo "02. GET /v2/entities"
```
Comment (fgalan) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity.test]: GET /v2/entities -> GET /v2/entities/8787GHY

File: test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity.test [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+            "statusCode": {
+                "code": "200",
+                "reasonPhrase": "OK"
+            }
+        }
+    ]
+}
+
+
+02. GET /v2/entities
```
Comment (fgalan) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity.test]: GET /v2/entities -> GET /v2/entities/8787GHY

File: test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+    }
+  ],
+  "updateAction": "APPEND"
+}'
+orionCurl --url /v1/updateContext --payload "$payload" --json
+echo
+echo
+
+
+echo "03. GET /v2/entities"
```
Comment (fgalan) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test]: GET /v2/entities -> GET /v2/entities/E1

File: test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+    }
+  ],
+  "updateAction": "APPEND"
+}'
+orionCurl --url /v1/updateContext --payload "$payload" --json
+echo
+echo
+
+
+echo "05. GET /v2/entities"
```
Comment (fgalan) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test]: GET /v2/entities -> GET /v2/entities/E2

File: test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_not_found.test [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+
+#
+# 01. GET /v2/entities
+# 02. POST /v1/updateContext
+# 03. GET /v2/entities
+#
+
+echo "01. GET /v2/entities"
+echo "===================="
+orionCurl --url /v2/entities/E1 --json
```
Comment (fgalan) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_not_found.test]: GET /v2/entities -> GET /v2/entities/E1

File: test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_not_found.test [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+    }
+  ],
+  "updateAction": "APPEND"
+}'
+orionCurl --url /v1/updateContext --payload "$payload" --json
+echo
+echo
+
+
+echo "03. GET /v2/entities"
```
Comment (fgalan) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_not_found.test]: GET /v2/entities -> GET /v2/entities/E1

File: test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+dbInit CB
+brokerStart CB
+
+--SHELL--
+
+#
+# 01. GET /v2/entities
+# 02. POST /v1/updateContext
+# 03. GET /v2/entities
+# 04. GET /v2/entities?count=true
```
Comment (jmcanterafonseca) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test]: ?options=count as per the latest spec

File: test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+--REGEXPECT--
+01. GET /v2/entities
+====================
+HTTP/1.1 404 Not Found
+Content-Length: 36
+Content-Type: application/json
+Date: REGEX(.*)
+
+{
+    "error": "No context element found"
```
Comment (jmcanterafonseca) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test]: description: is missing

Comment (jmcanterafonseca) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test]: As a side note and this is for the spec discussion but perhaps error can be a tabulated string such as 'NotFound' , following a similar approach than DOMError.name https://developer.mozilla.org/en-US/docs/Web/API/DOMError

I think this approach was suggested in the spec although not formalized yet

Comment (fgalan) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test]: Description is an optional field, thus this error payload is ok from a "formal" point of view. However, if we'd add description, which exact message should be used here?

File: test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test (hunk 9 above)
Comment (fgalan) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test]: Probably is a typo in the test preamble comment, as step 4 below is 

"04. POST /v1/updateContext with compound attribute value"

Actually, this PR is on an operation which doesn't uses count...

File: test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test [hunk 11]
```diff
@@ -0,0 +1,356 @@
...
+
+echo "05. GET /v2/entities"
+echo "===================="
+orionCurl --url /v2/entities/E2 --json
+echo
+echo
+
+
+--REGEXPECT--
+01. GET /v2/entities
```
Comment (jmcanterafonseca) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test]: check the descriptions as per the comment from Fermin

File: test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test [hunk 12]
```diff
@@ -0,0 +1,356 @@
...
+                "code": "200",
+                "reasonPhrase": "OK"
+            }
+        }
+    ]
+}
+
+
+03. GET /v2/entities
+====================
```
Comment (jmcanterafonseca) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test]: description

File: test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test [hunk 13]
```diff
@@ -0,0 +1,356 @@
...
+                "code": "200",
+                "reasonPhrase": "OK"
+            }
+        }
+    ]
+}
+
+
+05. GET /v2/entities
+====================
```
Comment (jmcanterafonseca) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test]: description

File: test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_not_found.test [hunk 14]
```diff
@@ -0,0 +1,277 @@
...
+
+--REGEXPECT--
+01. GET /v2/entities
+====================
+HTTP/1.1 404 Not Found
+Content-Length: 36
+Content-Type: application/json
+Date: REGEX(.*)
+
+{
```
Comment (jmcanterafonseca) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_not_found.test]: description missing

File: test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test [hunk 15]
```diff
@@ -0,0 +1,356 @@
+# Copyright 2015 Telefonica Investigacion y Desarrollo, S.A.U
```
Comment (jmcanterafonseca) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test]: I'm wondering whether this test is really needed as entity rendering has already been tested by /v2/entities tests

File: test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_not_found.test [hunk 16]
```diff
@@ -0,0 +1,277 @@
+# Copyright 2015 Telefonica Investigacion y Desarrollo, S.A.U
```
Comment (jmcanterafonseca) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_not_found.test]: The former test was already testing for non existence. I think it would be better not to repeat tests ... or focus them in testing only one thing

File: test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_too_many.test [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+
+02. GET /v2/entities/E1
+=============================
+HTTP/1.1 409 Conflict
+Content-Length: 73
+Content-Type: application/json
+Date: REGEX(.*)
+
+{
+    "description": "/v2/entities?id=E1",
```
Comment (jmcanterafonseca) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_too_many.test]: The description should be more descriptive. A message like 'There is more than one entity with E1 as Id. you should refine your query with a type or you can list all the entities invoking ...' but that should be discussed as part of the spec work I think

File: src/app/contextBroker/contextBroker.cpp [hunk 18]
```diff
@@ -346,6 +347,11 @@ PaArgument paArgs[] =
 #define ENT_COMPS_V2       2, { "v2", "entities" }
 #define ENT_COMPS_WORD     ""
 
+
+#define IENT                EntityRequest
+#define IENT_COMPS_V2       3, { "v2", "entities", "*" }
+#define IENT_COMPS_WORD     ""
```
Comment (jmcanterafonseca) [inline src/app/contextBroker/contextBroker.cpp]: what does I stand for?

File: src/lib/rest/HttpStatusCode.h [hunk 19]
```diff
@@ -52,7 +53,7 @@ typedef enum HttpStatusCode
   SccEntityTypeRequired     = 481,   // The EntityType is required by the receiver
   SccAttributeListRequired  = 482,   // The Attribute List is required by the receiver
   SccReceiverInternalError  = 500,   // An unknown error at the receiver has occurred
-  SccNotImplemented         = 501    // The given operation is not implemented
+  SccNotImplemented         = 501   // The given operation is not implemented
```
Comment (jmcanterafonseca) [inline src/lib/rest/HttpStatusCode.h]: nit: Is that change needed?

File: src/lib/serviceRoutinesV2/getEntity.cpp [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero
+* General Public License for more details.
+*
+* You should have received a copy of the GNU Affero General Public License
+* along with Orion Context Broker. If not, see http://www.gnu.org/licenses/.
+*
+* For those usages not covered by this license please contact with
+* iot_support at tid dot es
+*
+* Author: Ken Zangelin
```
Comment (jmcanterafonseca) [inline src/lib/serviceRoutinesV2/getEntity.cpp]: nit: do you want to give credits to Ken? ;)

File: src/lib/serviceRoutinesV2/getEntity.h [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero
+* General Public License for more details.
+*
+* You should have received a copy of the GNU Affero General Public License
+* along with Orion Context Broker. If not, see http://www.gnu.org/licenses/.
+*
+* For those usages not covered by this license please contact with
+* iot_support at tid dot es
+*
+* Author: Ken Zangelin
```
Comment (jmcanterafonseca) [inline src/lib/serviceRoutinesV2/getEntity.h]: same

**Comment by jmcanterafonseca** [comment]
@crbrox good work. I'm happy to see that is not difficult to adapt the existing code base to the new API ;)

File: src/app/contextBroker/contextBroker.cpp (hunk 18 above)
Comment (crbrox) [inline src/app/contextBroker/contextBroker.cpp]: I - Individual entity , named like https://github.com/telefonicaid/fiware-orion/blob/develop/src/app/contextBroker/contextBroker.cpp#L468 and following ones

File: src/lib/serviceRoutinesV2/getEntity.cpp (hunk 20 above)
Comment (crbrox) [inline src/lib/serviceRoutinesV2/getEntity.cpp]: fixed 215d1996010e62a58547424dde3dc972fdcf9afb

File: src/lib/serviceRoutinesV2/getEntity.h (hunk 21 above)
Comment (crbrox) [inline src/lib/serviceRoutinesV2/getEntity.h]: fixed 215d1996010e62a58547424dde3dc972fdcf9afb

File: src/lib/rest/HttpStatusCode.h (hunk 19 above)
Comment (crbrox) [inline src/lib/rest/HttpStatusCode.h]: Fixed df05e06d8bdc64b95ea7dc4e25765f25f86ff5a8

File: src/lib/apiTypesV2/Entity.cpp (hunk 1 above)
Comment (crbrox) [inline src/lib/apiTypesV2/Entity.cpp]: Fixed 82ea64d7dafa380e05b48c99db4d74ae0e6beac9

File: test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test (hunk 15 above)
Comment (crbrox) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test]: Fixed 33f4d110adc1b6ca0f9ac1013dd698532f446f59 (removed)

File: test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test (hunk 13 above)
Comment (crbrox) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test]: Removed at 33f4d110adc1b6ca0f9ac1013dd698532f446f59

File: test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test (hunk 12 above)
Comment (crbrox) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test]: Removed at 33f4d110adc1b6ca0f9ac1013dd698532f446f59

File: test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity.test (hunk 2 above)
Comment (fgalan) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity.test]: Fixed in c59b819

File: test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity.test (hunk 3 above)
Comment (fgalan) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity.test]: Fixed in c59b819

File: test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity.test (hunk 4 above)
Comment (fgalan) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity.test]: Fixed in c59b819

File: test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test (hunk 5 above)
Comment (fgalan) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test]: Fixed in c59b819

File: test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test (hunk 6 above)
Comment (fgalan) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test]: Fixed in c59b819

File: test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_not_found.test (hunk 7 above)
Comment (fgalan) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_not_found.test]: Fixed in c59b819

File: test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_not_found.test (hunk 8 above)
Comment (fgalan) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_not_found.test]: Fixed in c59b819

File: test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test (hunk 9 above)
Comment (fgalan) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test]: Finally (as in c59b819) the `count=true` is not shown in the .test file. Fixed.

File: test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test (hunk 10 above)
Comment (fgalan) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test]: Currently, these error messages are generated by the `httpStatusCodeString(HttpStatusCode code)` function in HttpStatusCode.cpp. Thus, if we want DOMError-style names for v2, it could be easy to provide a similar version to do a new code-to-string mapping. @jmcanterafonseca would you like to do that? ;)

Regarding description, it is ok with me to have that test. I think is a matter of adjusting the render() for the new Entity class. The only problem would be to align this message with the one used by standard operations in the same situation (they would be different).

File: test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test (hunk 11 above)
Comment (fgalan) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_complex.test]: FIxed in  c59b819.

File: test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_not_found.test (hunk 14 above)
Comment (fgalan) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_not_found.test]: If you refer to .test step description, fixed in c59b819. If you refer to description field in error message, it is being discussed in another comment thread.

File: test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_not_found.test (hunk 16 above)
Comment (fgalan) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_not_found.test]: Not sure if understanding this comment... maybe this comment is obsolete at c59b819, as I think all the test are ok and their steps make sense. However, if I'm wrong, please drop another comment about the problem.

File: test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_too_many.test (hunk 17 above)
Comment (fgalan) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_too_many.test]: I would suggest:

"There is more than one entity with E1 as Id, you can get the list of all the entities invoking: GET /v2/entities?id=E1"

I think that suggesting refining with type is not a "complete" hint: even adding type it could be a conflict (due to service paths), so probably it is better not trying to be too much clever from our side :)

Do you really think spec should include the exact text of descriptions? I think that error messages (e.g. "NotFound, etc.) should be included in the spec because client could make decission based on their values, but descriptions are text for humans and could be left free to API implementations.

File: src/lib/serviceRoutinesV2/getEntity.cpp [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+  // 02. Call standard op postQueryContext
+  postQueryContext(ciP, components, compV, parseDataP);
+
+
+  // 03. Render entity response
+  entity.fill(&parseDataP->qcrs.res);
+  answer = entity.render(ciP, EntityResponse);
+  if ( parseDataP->qcrs.res.errorCode.code == SccOk &&
+       parseDataP->qcrs.res.contextElementResponseVector.size()>1)
+  {
```
Comment (kzangeli) [inline src/lib/serviceRoutinesV2/getEntity.cpp]: The style used in the broker is the following:

    if ((parseDataP->qcrs.res.errorCode.code == SccOk) && (parseDataP->qcrs.res.contextElementResponseVector.size() > 1))

1. No space after ')'
2. Space before and after operatoros
3. Conditions within parenthesis if more than one condition
4. If the line is shorter than 120 chars, use a single line

File: src/lib/serviceRoutinesV2/getEntity.cpp [hunk 23]
```diff
@@ -0,0 +1,93 @@
...
+std::string getEntity
+(
+  ConnectionInfo*            ciP,
+  int                        components,
+  std::vector<std::string>&  compV,
+  ParseData*                 parseDataP
+)
+{
+  std::string  answer;
+  Entity     entity;
```
Comment (kzangeli) [inline src/lib/serviceRoutinesV2/getEntity.cpp]: alignment of variable names (add two spaces before 'entity')

File: src/lib/serviceRoutinesV2/getEntity.h [hunk 24]
```diff
@@ -0,0 +1,49 @@
+#ifndef SRC_LIB_SERVICEROUTINES_GETENTITYV2_H_
+#define SRC_LIB_SERVICEROUTINES_GETENTITYV2_H_
```
Comment (kzangeli) [inline src/lib/serviceRoutinesV2/getEntity.h]: The name to use for the define is:

    SRC_LIB_SERVICEROUTINESV2_GETENTITY_H_

(in three places)

File: test/functionalTest/cases/0950_GET_v2_entity/GET_v2_entity.test [hunk 25]
```diff
@@ -0,0 +1,159 @@
...
+
+--SHELL-INIT--
+dbInit CB
+brokerStart CB
+
+--SHELL--
+
+#
+# 01. POST /v1/updateContext to create 8787GHY entity
+# 03. GET /v2/entities/8787GHY
```
Comment (kzangeli) [inline test/functionalTest/cases/0950_GET_v2_entity/GET_v2_entity.test]: 02

File: test/functionalTest/cases/0950_GET_v2_entity/GET_v2_entity_not_found.test [hunk 26]
```diff
@@ -0,0 +1,277 @@
...
+# along with Orion Context Broker. If not, see http://www.gnu.org/licenses/.
+#
+# For those usages not covered by this license please contact with
+# iot_support at tid dot es
+
+# VALGRIND_READY - to mark the test ready for valgrindTestSuite.sh
+
+--NAME--
+GET Entity V2 Not found
+--SHELL-INIT--
```
Comment (kzangeli) [inline test/functionalTest/cases/0950_GET_v2_entity/GET_v2_entity_not_found.test]: Empty line missing before --SHELL-INIT--   (not strictly necessary but it makes it easier to read and especially, all other test files (except perhaps very old ones) are made that way)

File: src/app/contextBroker/contextBroker.cpp (hunk 18 above)
Comment (jmcanterafonseca) [inline src/app/contextBroker/contextBroker.cpp]: NTC

File: test/functionalTest/cases/0950_GET_v2_entity/GET_v2_entity_not_found.test [hunk 27]
```diff
@@ -0,0 +1,277 @@
...
+--REGEXPECT--
+01. GET /v2/entities/E1 (error)
+===============================
+HTTP/1.1 404 Not Found
+Content-Length: 36
+Content-Type: application/json
+Date: REGEX(.*)
+
+{
+    "error": "No context element found"
```
Comment (jmcanterafonseca) [inline test/functionalTest/cases/0950_GET_v2_entity/GET_v2_entity_not_found.test]: it should be 

```js
error: "NotFound",
description: "The requested entity has not been found. Check type and id"
```

File: test/functionalTest/cases/0950_GET_v2_entity/GET_v2_entity_too_many.test [hunk 28]
```diff
@@ -0,0 +1,159 @@
...
+02. GET /v2/entities/E1 (too many entities error)
+=================================================
+HTTP/1.1 409 Conflict
+Content-Length: 73
+Content-Type: application/json
+Date: REGEX(.*)
+
+{
+    "description": "/v2/entities?id=E1",
+    "error": "Many entities with that ID"
```
Comment (jmcanterafonseca) [inline test/functionalTest/cases/0950_GET_v2_entity/GET_v2_entity_too_many.test]: The description should be clearer. "There is more than one entity with id "E1". Refine your query by providing a type"

File: test/functionalTest/cases/0950_GET_v2_entity/GET_v2_entity_too_many.test [hunk 29]
```diff
@@ -0,0 +1,159 @@
...
+
+02. GET /v2/entities/E1 (too many entities error)
+=================================================
+HTTP/1.1 409 Conflict
+Content-Length: 73
+Content-Type: application/json
+Date: REGEX(.*)
+
+{
+    "description": "/v2/entities?id=E1",
```
Comment (jmcanterafonseca) [inline test/functionalTest/cases/0950_GET_v2_entity/GET_v2_entity_too_many.test]: The description should be clearer. "There is more than one entity with id "E1". Refine your query by providing a type"

Comment (fgalan) [inline test/functionalTest/cases/0950_GET_v2_entity/GET_v2_entity_too_many.test]: "Refine your query by providing a type" could be missleading: adding a type couldn't solve the problem if "duplicity" is due to service path. Thus, let's keep it with "Theres is more than one entity with id X. Refine your query" (without mentionint type, as it is only one way of refining, but not the only one and not necessarily the one that would provide the solution).

File: src/lib/serviceRoutinesV2/getEntity.h (hunk 24 above)
Comment (crbrox) [inline src/lib/serviceRoutinesV2/getEntity.h]: fixed in f18c358

File: src/lib/serviceRoutinesV2/getEntity.cpp (hunk 23 above)
Comment (crbrox) [inline src/lib/serviceRoutinesV2/getEntity.cpp]: fixed in 1881ada

File: src/lib/serviceRoutinesV2/getEntity.cpp (hunk 22 above)
Comment (crbrox) [inline src/lib/serviceRoutinesV2/getEntity.cpp]: fixed in 1881ada

File: test/functionalTest/cases/0950_GET_v2_entity/GET_v2_entity.test (hunk 25 above)
Comment (crbrox) [inline test/functionalTest/cases/0950_GET_v2_entity/GET_v2_entity.test]: fixed in e45787b

File: test/functionalTest/cases/0950_GET_v2_entity/GET_v2_entity_not_found.test (hunk 26 above)
Comment (crbrox) [inline test/functionalTest/cases/0950_GET_v2_entity/GET_v2_entity_not_found.test]: Fixed in d3ba7def

File: src/lib/apiTypesV2/Entity.cpp [hunk 30]
```diff
@@ -128,7 +136,30 @@ void Entity::fill(const std::string& _id, const std::string& _type, const std::s
...
 
+  if (qcrsP->errorCode.code != SccOk)
+  {
+    //
+    // If no entity has been found, we respond with a 404 NOT FOUND
+    // If any other error - use the error for the response
+    //
+    errorCode.fill(qcrsP->errorCode);
+  }
+  else if (qcrsP->contextElementResponseVector.size()>1)
```
Comment (kzangeli) [inline src/lib/apiTypesV2/Entity.cpp]: spaces around '>'

File: src/lib/apiTypesV2/Entity.cpp [hunk 31]
```diff
@@ -128,7 +136,30 @@ void Entity::fill(const std::string& _id, const std::string& _type, const std::s
...
+  {
+      //
+      // If there are more than one entity, we return an error
+      //
+      errorCode.fill("Many entities with that ID", "/v2/entities?id="+qcrsP->contextElementResponseVector[0]->contextElement.entityId.id);
+  }
+  else
+  {
+    ContextElement* ceP = &qcrsP->contextElementResponseVector[0]->contextElement;
+    this->fill(ceP->entityId.id, ceP->entityId.type, ceP->entityId.isPattern, &ceP->contextAttributeVector);
```
Comment (kzangeli) [inline src/lib/apiTypesV2/Entity.cpp]: 'this' should not be needed here

File: src/lib/apiTypesV2/Entity.cpp [hunk 32]
```diff
@@ -128,7 +136,30 @@ void Entity::fill(const std::string& _id, const std::string& _type, const std::s
...
+void Entity::fill(QueryContextResponse* qcrsP)
+{
 
+  if (qcrsP->errorCode.code != SccOk)
+  {
+    //
+    // If no entity has been found, we respond with a 404 NOT FOUND
+    // If any other error - use the error for the response
+    //
+    errorCode.fill(qcrsP->errorCode);
```
Comment (kzangeli) [inline src/lib/apiTypesV2/Entity.cpp]: This comment is a little bit strange, makes me think that code is missing.
Perhaps the comment should be outside 'if (qcrsP->errorCode.code != SccOk)' ?

File: src/lib/apiTypesV2/ErrorCode.cpp [hunk 33]
```diff
@@ -85,7 +85,7 @@ std::string ErrorCode::toJson(bool isLastElement)
   out += "{";
 
   bool comma = false;
-  if ((error != "") && (error == "OK"))
+  if ((error != "") && (error != "OK"))
```
Comment (kzangeli) [inline src/lib/apiTypesV2/ErrorCode.cpp]: Good! :-D

File: src/lib/common/statistics.cpp [hunk 34]
```diff
@@ -211,5 +214,8 @@ void statisticsUpdate(RequestType request, Format inFormat)
   case EntitiesResponse:                                 ++noOfEntitiesResponses; break;
   case EntryPointsRequest:                               ++noOfEntryPointsRequests; break;
   case EntryPointsResponse:                              ++noOfEntryPointsResponses; break;
+  case EntityRequest:                                    ++noOfEntityRequests; break;
+  case EntityResponse:                                   ++noOfEntityResponses; break;
+
```
Comment (kzangeli) [inline src/lib/common/statistics.cpp]: The counters should be added to the output of 'statisticsTreat' also (in serviceRoutines)

Comment (crbrox) [inline src/lib/common/statistics.cpp]: I think that was meant for a further issue improving functionality.  AFAICS,  /v2/entities does not implement it yet, likely for the same reason ... To be confirmed by @fgalan

Comment (kzangeli) [inline src/lib/common/statistics.cpp]: You are probably right.
We need a real fix for the statistics, probably in a set of new PRs.
Some day ...

NTC

Comment (fgalan) [inline src/lib/common/statistics.cpp]: Regarding statistics improvement, the following issue is opened: https://github.com/telefonicaid/fiware-orion/issues/27. However, we shouldn't address it until PO defines the exact requirements for that operation.

File: src/lib/apiTypesV2/Entity.cpp (hunk 32 above)
Comment (crbrox) [inline src/lib/apiTypesV2/Entity.cpp]: Fixed in c328529

File: src/lib/apiTypesV2/Entity.cpp (hunk 31 above)
Comment (crbrox) [inline src/lib/apiTypesV2/Entity.cpp]: Fixed in c328529

File: src/lib/apiTypesV2/Entity.cpp (hunk 30 above)
Comment (crbrox) [inline src/lib/apiTypesV2/Entity.cpp]: Fixed in c328529

File: test/functionalTest/cases/0950_GET_v2_entity/GET_v2_entity_too_many.test (hunk 29 above)
Comment (crbrox) [inline test/functionalTest/cases/0950_GET_v2_entity/GET_v2_entity_too_many.test]: Fixed in c328529

File: test/functionalTest/cases/0950_GET_v2_entity/GET_v2_entity_too_many.test (hunk 28 above)
Comment (crbrox) [inline test/functionalTest/cases/0950_GET_v2_entity/GET_v2_entity_too_many.test]: Fixed in c328529

File: test/functionalTest/cases/0950_GET_v2_entity/GET_v2_entity_not_found.test (hunk 27 above)
Comment (crbrox) [inline test/functionalTest/cases/0950_GET_v2_entity/GET_v2_entity_not_found.test]: Fixed in c328529

File: test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_not_found.test (hunk 16 above)
Comment (fgalan) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_not_found.test]: No coment so far, so I understand NTC.

File: test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_too_many.test (hunk 17 above)
Comment (fgalan) [inline test/functionalTest/cases/950_GET_v2_entity/GET_v2_entity_too_many.test]: NTC (I guess)

**Comment by fgalan** [comment]
LGTM (passing the ball to @jmcanterafonseca for LGTM)

**Comment by jmcanterafonseca** [comment]
LGTM nice work @crbrox
